src/cio_websocket.c

```diff
@@ -0,0 +1,466 @@
+	cio_write_buffer_element_init(&s->wb_send_payload, payload, length);
+	cio_write_buffer_queue_tail(&s->wbh, &s->wb_send_header);
+	cio_write_buffer_queue_tail(&s->wbh, &s->wb_send_payload);
+
+	s->bs->write(s->bs, &s->wbh, written_cb, s);
+}
+
+static void close(struct cio_websocket *ws)
+{
+	if (ws->close_hook) {
+		ws->close_hook(ws);
+	}
+}
+
+static bool is_status_code_invalid(uint16_t status_code)
+{
+	if ((status_code >= CIO_WEBSOCKET_CLOSE_NORMAL) && (status_code <= CIO_WEBSOCKET_CLOSE_UNSUPPORTED)) {
+		return false;
+	}
+
+	if ((status_code >= CIO_WEBSOCKET_CLOSE_UNSUPPORTED_DATA) && (status_code <= CIO_WEBSOCKET_CLOSE_INTERNAL_ERROR)) {
+		return false;
+	}
+
+	if ((status_code >= CIO_WEBSOCKET_CLOSE_RESERVED_LOWER_BOUND) && (status_code <= CIO_WEBSOCKET_CLOSE_RESERVED_UPPER_BOUND)) {
+		return false;
+	}
+
+	return true;
+}
+
+static void close_frame_written(struct cio_buffered_stream *bs, void *handler_context, const struct cio_write_buffer *buffer, enum cio_error err)
+{
+	(void)bs;
+	(void)buffer;
+	(void)err;
+	struct cio_websocket *ws = (struct cio_websocket *)handler_context;
+	if (ws->self_initiated_close) {
+		receive_frames(ws);
+	} else {
+		close(ws);
+	}
+}
+
+static void do_nothing(struct cio_buffered_stream *bs, void *handler_context, const struct cio_write_buffer *buffer, enum cio_error err)
+{
+	(void)bs;
+	(void)handler_context;
+	(void)buffer;
+	(void)err;
+}
+
+static void close_timeout_handler(struct cio_timer *timer, void *handler_context, enum cio_error err)
+{
+	(void)timer;
+	(void)err;
+	struct cio_websocket *ws = (struct cio_websocket *)handler_context;
+	close(ws);
+}
+
+static void send_close_frame(struct cio_websocket *ws, enum cio_websocket_status_code status_code)
+{
+	ws->close_status = cio_htobe16(status_code);
+	enum cio_error err = cio_timer_init(&ws->close_timer, ws->loop, NULL);
+	if (err == CIO_SUCCESS) {
+		err = ws->close_timer.expires_from_now(&ws->close_timer, close_timeout_ns, close_timeout_handler, ws);
+		if (err == CIO_SUCCESS) {
+			send_frame(ws, (uint8_t *)&ws->close_status, sizeof(ws->close_status), CIO_WEBSOCKET_CLOSE_FRAME, close_frame_written);
+			return;
+		}
+	}
+
+	send_frame(ws, (uint8_t *)&ws->close_status, sizeof(ws->close_status), CIO_WEBSOCKET_CLOSE_FRAME, do_nothing);
+	close(ws);
+}
+
+static void handle_frame(struct cio_websocket *ws, uint8_t *data, uint64_t length)
+{
+	if (unlikely(ws->is_server && (ws->ws_flags.mask == 0))) {
+		// TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		return;
+	}
+
+	if (ws->ws_flags.mask != 0) {
+		cio_websocket_mask(data, length, ws->mask);
+	}
+
+	if (unlikely((ws->ws_flags.fin == 0) && (ws->ws_flags.opcode >= CIO_WEBSOCKET_CLOSE_FRAME))) {
+		// TODO: handle_error(s, WS_CLOSE_PROTOCOL_ERROR);
+		return;
+	}
+
+	if (ws->ws_flags.fin == 0) {
+		if (ws->ws_flags.opcode != CIO_WEBSOCKET_CONTINUATION_FRAME) {
+			if (unlikely(ws->ws_flags.is_fragmented)) {
+				// TODO: log_err("Overwriting Opcode of unfinished fragmentation!");
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this block of commented out code. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3ACommentedCode)
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS1135)